### PR TITLE
gh-64490: Fix bugs in argument clinic varargs processing

### PR DESCRIPTION
**Pull request by colorfulappl**
Fix 3 bugs introduced in #18609.

Bug reason lists in each commit's commit message.

https://bugs.python.org/issue20291

* Issue: gh-64490

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -941,13 +941,15 @@ def parser_body(prototype, *fields, declarations=''):
                     max_pos,
                     min_kw_only
                 )
+                pre_buffer = "nargs"
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Can we name this something more descriptive than `pre_buffer`? For example simply `nargs`, unless it is taken (or `nargs_str`)?

**Comment by erlend-aasland** [comment]
Could you also add a NEWS entry?

**Comment by erlend-aasland** [comment]
Thanks; still missing the NEWS entry, though.

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (colorfulappl) [reply Tools/clinic/clinic.py]: It's renamed to `nargs` now.

**Comment by erlend-aasland** [comment]
> NEWS added :)

Thanks; that's some entry! :) Could you narrow it down to a couple of concise sentences that describe the actual changes only?

**Comment by erlend-aasland** [comment]
Also, could you please add tests for each case?

**Comment by colorfulappl** [comment]
> Also, could you please add tests for each case?

Do you mean put the test cases into test_linic.py?

To see if the OOB bug is fixed in commit https://github.com/python/cpython/pull/32092/commits/620396279cf94e09bd7cd6e925e6e80471130d6f , I should write some code snippet calling function `_PyArg_UnpackKeywordsWithVararg`, compile it and ensure that it does not crash.

I'm not sure how can I do this without touching the Python vm or library,
though I have wrote a proof-of-concept by adding a built-in function and test it manually:
https://github.com/colorfulappl/cpython/tree/unpack_keywords_oob_bug_poc .


As for commit https://github.com/python/cpython/pull/32092/commits/6793f3838077e3cf12d4e44059396478757522b2, I'm looking at test_linic.py and will have a try later.

File: Misc/NEWS.d/next/Tools-Demos/2022-08-11-09-58-15.gh-issue-64490.PjwhM4.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Fix an out-of-bound bug of :c:func:`_PyArg_UnpackKeywordsWithVararg`.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Tools-Demos/2022-08-11-09-58-15.gh-issue-64490.PjwhM4.rst]:
```suggestion
Fix an out-of-bound bug of ``_PyArg_UnpackKeywordsWithVararg``.
```
It is undocumented so linking will not work.

Comment (erlend-aasland) [reply Misc/NEWS.d/next/Tools-Demos/2022-08-11-09-58-15.gh-issue-64490.PjwhM4.rst]: > It is undocumented so linking will not work.

True, but literals are rendered differently from functions. I would instead do this:

```suggestion
Fix an out-of-bound bug of :c:func:`!_PyArg_UnpackKeywordsWithVararg`.
```

Also, I think the NEWS entry needs a little bit more tweaking before landing this.

**Review by kumaraditya303 — COMMENTED** [review]
LGTM, I verified it with your tests and it works correctly [^1]. I think we should have functional tests for AC but currently there aren't any so that's a separate issue.

[^1]: https://github.com/colorfulappl/cpython/tree/unpack_keywords_oob_bug_poc

**Review by kumaraditya303 — APPROVED** [review]

**Comment by erlend-aasland** [comment]
> I think we should have functional tests for AC but currently there aren't any so that's a separate issue.

+1

Would you mind opening an issue[^1]?

[^1]: assuming there is none open already]

**Comment by erlend-aasland** [comment]
I took the liberty of editing the NEWS entry. Leave a thumbs-up if it looks ok to you.

I'll land this in a couple of days. Kumar, do you think Eric is interested in reviewing this? I guess he is busy with other PRs and PEPs, so I'll refrain from pinging him.

**Comment by kumaraditya303** [comment]
> Would you mind opening an issue[1](https://github.com/python/cpython/pull/32092#user-content-fn-1-7d9e1473f9f9970e831f8df181d89538)?

Created issue https://github.com/python/cpython/issues/96002

**Comment by kumaraditya303** [comment]
> I'll land this in a couple of days. Kumar, do you think Eric is interested in reviewing this? I guess he is busy with other PRs and PEPs, so I'll refrain from pinging him.

Ya, he is busy with other work and PEP. This is small so I think safe to land.

**Comment by erlend-aasland** [comment]
Ok, then I'll think we're ready for landing; I'll merge it later tonight.

**Comment by erlend-aasland** [comment]
I see that Batuhan self-requested a review; I'll give him a couple of more days to do it before landing. cc. @isidentical

**Review by isidentical — APPROVED** [review]
Thanks for the ping @erlend-aasland, and thanks for the PR @colorfulappl. I've gone through the changes, and they all look good to me. Quick question below:

File: Python/getargs.c [hunk 3]
```diff
@@ -2586,7 +2586,7 @@ _PyArg_UnpackKeywordsWithVararg(PyObject *const *args, Py_ssize_t nargs,
             current_arg = NULL;
         }
 
-        buf[i + vararg + 1] = current_arg;
+        buf[i + 1] = current_arg;
```
Comment (isidentical) [inline Python/getargs.c]: Is there a case where the problem in this change can be triggered?

Comment (colorfulappl) [reply Python/getargs.c]: >To see if the OOB bug is fixed in commit https://github.com/python/cpython/commit/620396279cf94e09bd7cd6e925e6e80471130d6f , I should write some code snippet calling function _PyArg_UnpackKeywordsWithVararg, compile it and ensure that it does not crash.
>
> I'm not sure how can I do this without touching the Python vm or library,
though I have wrote a proof-of-concept by adding a built-in function and test it manually:
https://github.com/colorfulappl/cpython/tree/unpack_keywords_oob_bug_poc .


I have written a PoC as a built-in function here:
https://github.com/colorfulappl/cpython/commit/8e7aa16b093fe027c9fd9a8e0f3f9b329cd507b3

You can build it and type
```python
oob_poc(1, 2, 3, 4, kw1=5, kw2=6)
```
to trigger this bug.

Comment (erlend-aasland) [reply Python/getargs.c]: @kumaraditya303, have you started working on #96002 yet? This test would be a good candidate. We might want to put this PR on hold until that is implemented. OTOH, we could just add the referenced test to the test suite in #96002.

Comment (kumaraditya303) [reply Python/getargs.c]: I haven't started yet, feel free to take over it.

Comment (erlend-aasland) [reply Python/getargs.c]: I don't have time for it right now; my schedule is full this weekend. Possibly next week, but I can't promise anything.

Comment (colorfulappl) [reply Python/getargs.c]: @erlend-aasland I'm willing to write a draft this weekend, if you don't mind.

Comment (erlend-aasland) [reply Python/getargs.c]: Sure, go ahead!

Comment (erlend-aasland) [reply Python/getargs.c]: > I'm willing to write a draft this weekend, if you don't mind.

FTR, the PR is up: gh-96178

**Comment by smontanaro** [comment]
(I'm working my way through some PRs which have been approved and are labeled "awaiting merge", hence my seemingly bolt from the blue comment. Why? [Read here](https://discuss.python.org/t/prs-awaiting-merge/21313).)

I looked through the comments. Is it still waiting for further changes?

**Comment by erlend-aasland** [comment]
> Is it still waiting for further changes?

Since gh-96178 has recently been merged, we finally have a test framework for functional clinic tests, so I'd say this PR should add some tests before we are ready to land.

**Comment by colorfulappl** [comment]
I added some test cases and did another change in [a48971d](https://github.com/python/cpython/pull/32092/commits/a48971dd8dc04e5712ff5bc78b20cf483c297d9d), so this PR may be necessary to be reviewed again. @erlend-aasland @isidentical @kumaraditya303

**Comment by colorfulappl** [comment]
Another change in e6a4e13:
Argument Clinic does not generate `noptargs` when there is a vararg and no optional argument.

File: Python/getargs.c (hunk 3 above)
Comment (erlend-aasland) [reply Python/getargs.c]: It looks to me like we can mark this conversation as resolved.

**Review by erlend-aasland — APPROVED** [review]
LGTM, thanks! Kumar, Batuhan, do you have further remarks?

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Comment by erlend-aasland** [comment]
@colorfulappl, please fix the merge conflicts

**Comment by erlend-aasland** [comment]
@kumaraditya303 I'm inclined to backport the functional tests for AC to 3.11 and 3.10. What do you think? We've backported tests to increase coverage in the maintained branches before; it would help backporting the AC fixes (and by the way, we should backport the AC bug fixes that we landed earlier today)

**Comment by colorfulappl** [comment]
I have backported this to 3.11 (#100368).
And since 3.10 Argument Clinic does not support varargs parsing, there is no need to backport this to 3.10, IMHO.

**Comment by serhiy-storchaka** [comment]
It is too later for 3.10.

**Pull request merged**